File: CareBooking/Models/UserManagementModels.swift
import Foundation

enum TeamAccessRole: String, CaseIterable, Identifiable, Hashable {
    case admin
    case collaborator
    case viewer

    var id: String { rawValue }

    var title: String {
        switch self {
        case .admin: "Admin"
        case .collaborator: "Collaborator"
        case .viewer: "Viewer"
        }
    }

    var displayTitle: String {
        switch self {
        case .admin: "Admin with Full access"
        case .collaborator, .viewer: title
        }
    }

    var canEditRole: Bool { self != .admin }

    var canExpandPermissions: Bool { self != .admin }

    var summaryBullets: [String] {
        switch self {
        case .admin:
            [
                "Admin can manage users, roles, and full workspace access.",
                "Admin can NOT be demoted by collaborators or viewers."
            ]
        case .collaborator:
            [
                "Collaborator can do what what what.",
                "Collaborator can NOT do what what what."
            ]
        case .viewer:
            [
                "Viewer can do what what what.",
                "Viewer can NOT do what what what."
            ]
        }
    }

    static var inviteChoices: [TeamAccessRole] {
        [.collaborator, .viewer]
    }
}

enum FeatureAccessKey: String, CaseIterable, Identifiable, Hashable {
    case profile
    case preference
    case booking
    case messaging
    case payment
    case notes
    case review

    var id: String { rawValue }

    var title: String {
        switch self {
        case .profile: "Profile feature access"
        case .preference: "Preference feature access"
        case .booking: "Booking/Reschedule feature access"
        case .messaging: "Messaging feature access"
        case .payment: "Payment feature access"
        case .notes: "Notes"
        case .review: "Review"
        }
    }

    var hasNestedItems: Bool {
        switch self {
        case .preference, .payment: true
        default: false
        }
    }

    static let nestedLabels = [
        "Address (create/ edit) - care recipient(s)",
        "Phone # (create/ edit) - care recipient(s)",
        "Email (create/ edit) - care recipient(s)"
    ]

    static func defaults(for role: TeamAccessRole) -> [FeatureAccessKey: Bool] {
        switch role {
        case .admin:
            Dictionary(uniqueKeysWithValues: allCases.map { ($0, true) })
        case .collaborator, .viewer:
            [
                .profile: true,
                .preference: false,
                .booking: false,
                .messaging: false,
                .payment: true,
                .notes: true,
                .review: true
            ]
        }
    }
}

struct NestedFeatureAccess: Identifiable, Hashable {
    var id: String
    var title: String
    var isEnabled: Bool
}

struct ManagedTeamUser: Identifiable, Hashable {
    var id: String
    var firstName: String
    var lastName: String
    var email: String
    var role: TeamAccessRole
    var invitePending: Bool
    var permissions: [FeatureAccessKey: Bool]
    var nestedPermissions: [FeatureAccessKey: [NestedFeatureAccess]]

    var displayName: String {
        let full = "\(firstName) \(lastName)".trimmingCharacters(in: .whitespaces)
        return full.isEmpty ? firstName : firstName
    }

    var initials: String {
        let first = firstName.first.map(String.init) ?? ""
        let last = lastName.first.map(String.init) ?? ""
        let value = (first + last).uppercased()
        return value.isEmpty ? "?" : value
    }

    static func nestedDefaults(enabled: Bool = true) -> [FeatureAccessKey: [NestedFeatureAccess]] {
        Dictionary(uniqueKeysWithValues: FeatureAccessKey.allCases.compactMap { key in
            guard key.hasNestedItems else { return nil }
            let items = FeatureAccessKey.nestedLabels.enumerated().map { index, title in
                NestedFeatureAccess(
                    id: "\(key.rawValue)-\(index)",
                    title: title,
                    isEnabled: enabled
                )
            }
            return (key, items)
        })
    }

    static let samples: [ManagedTeamUser] = [
        ManagedTeamUser(
            id: "chi",
            firstName: "Chi",
            lastName: "Roger",
            email: "chi@carebooking.app",
            role: .admin,
            invitePending: false,
            permissions: FeatureAccessKey.defaults(for: .admin),
            nestedPermissions: nestedDefaults()
        ),
        ManagedTeamUser(
            id: "john",
            firstName: "John",
            lastName: "Smith",
            email: "john@carebooking.app",
            role: .collaborator,
            invitePending: false,
            permissions: FeatureAccessKey.defaults(for: .collaborator),
            nestedPermissions: nestedDefaults()
        ),
        ManagedTeamUser(
            id: "sharee",
            firstName: "Sharee",
            lastName: "Lee",
            email: "sharee@carebooking.app",
            role: .collaborator,
            invitePending: false,
            permissions: FeatureAccessKey.defaults(for: .collaborator),
            nestedPermissions: nestedDefaults()
        ),
        ManagedTeamUser(
            id: "matt",
            firstName: "Matt",
            lastName: "Nguyen",
            email: "matt@carebooking.app",
            role: .collaborator,
            invitePending: false,
            permissions: FeatureAccessKey.defaults(for: .collaborator),
            nestedPermissions: nestedDefaults()
        ),
        ManagedTeamUser(
            id: "toby",
            firstName: "Toby",
            lastName: "Chen",
            email: "toby@carebooking.app",
            role: .viewer,
            invitePending: false,
            permissions: FeatureAccessKey.defaults(for: .viewer),
            nestedPermissions: nestedDefaults()
        )
    ]
}

File: android/app/src/main/java/com/embelife/app/model/UserManagementModels.kt
package com.embelife.app.model

enum class TeamAccessRole(val title: String, val displayTitle: String) {
    Admin("Admin", "Admin with Full access"),
    Collaborator("Collaborator", "Collaborator"),
    Viewer("Viewer", "Viewer");

    val canEditRole: Boolean get() = this != Admin
    val canExpandPermissions: Boolean get() = this != Admin

    val summaryBullets: List<String>
        get() = when (this) {
            Admin -> listOf(
                "Admin can manage users, roles, and full workspace access.",
                "Admin can NOT be demoted by collaborators or viewers.",
            )
            Collaborator -> listOf(
                "Collaborator can do what what what.",
                "Collaborator can NOT do what what what.",
            )
            Viewer -> listOf(
                "Viewer can do what what what.",
                "Viewer can NOT do what what what.",
            )
        }

    companion object {
        val inviteChoices = listOf(Collaborator, Viewer)
    }
}

enum class FeatureAccessKey(val title: String, val hasNestedItems: Boolean = false) {
    Profile("Profile feature access"),
    Preference("Preference feature access", hasNestedItems = true),
    Booking("Booking/Reschedule feature access"),
    Messaging("Messaging feature access"),
    Payment("Payment feature access", hasNestedItems = true),
    Notes("Notes"),
    Review("Review");

    companion object {
        val nestedLabels = listOf(
            "Address (create/ edit) - care recipient(s)",
            "Phone # (create/ edit) - care recipient(s)",
            "Email (create/ edit) - care recipient(s)",
        )

        fun defaults(forRole: TeamAccessRole): Map<FeatureAccessKey, Boolean> = when (forRole) {
            TeamAccessRole.Admin -> entries.associateWith { true }
            TeamAccessRole.Collaborator, TeamAccessRole.Viewer -> mapOf(
                Profile to true,
                Preference to false,
                Booking to false,
                Messaging to false,
                Payment to true,
                Notes to true,
                Review to true,
            )
        }
    }
}

data class NestedFeatureAccess(
    val id: String,
    val title: String,
    var isEnabled: Boolean,
)

data class ManagedTeamUser(
    val id: String,
    var firstName: String,
    var lastName: String,
    var email: String,
    var role: TeamAccessRole,
    var invitePending: Boolean,
    var permissions: Map<FeatureAccessKey, Boolean>,
    var nestedPermissions: Map<FeatureAccessKey, List<NestedFeatureAccess>>,
) {
    val displayName: String
        get() = firstName.ifEmpty { "$firstName $lastName".trim() }

    val initials: String
        get() {
            val first = firstName.firstOrNull()?.toString().orEmpty()
            val last = lastName.firstOrNull()?.toString().orEmpty()
            return (first + last).uppercase().ifEmpty { "?" }
        }

    companion object {
        fun nestedDefaults(enabled: Boolean = true): Map<FeatureAccessKey, List<NestedFeatureAccess>> =
            FeatureAccessKey.entries.filter { it.hasNestedItems }.associateWith { key ->
                FeatureAccessKey.nestedLabels.mapIndexed { index, title ->
                    NestedFeatureAccess(
                        id = "${key.name.lowercase()}-$index",
                        title = title,
                        isEnabled = enabled,
                    )
                }
            }

        val samples: List<ManagedTeamUser> = listOf(
            ManagedTeamUser("chi", "Chi", "Roger", "chi@embelife.app", TeamAccessRole.Admin, false, FeatureAccessKey.defaults(TeamAccessRole.Admin), nestedDefaults()),
            ManagedTeamUser("john", "John", "Smith", "john@embelife.app", TeamAccessRole.Collaborator, false, FeatureAccessKey.defaults(TeamAccessRole.Collaborator), nestedDefaults()),
            ManagedTeamUser("sharee", "Sharee", "Lee", "sharee@embelife.app", TeamAccessRole.Collaborator, false, FeatureAccessKey.defaults(TeamAccessRole.Collaborator), nestedDefaults()),
            ManagedTeamUser("matt", "Matt", "Nguyen", "matt@embelife.app", TeamAccessRole.Collaborator, false, FeatureAccessKey.defaults(TeamAccessRole.Collaborator), nestedDefaults()),
            ManagedTeamUser("toby", "Toby", "Chen", "toby@embelife.app", TeamAccessRole.Viewer, false, FeatureAccessKey.defaults(TeamAccessRole.Viewer), nestedDefaults()),
        )
    }
}
